src/Reflex/EventWriter.hs

```diff
@@ -0,0 +1,172 @@
+-- | This module provides 'EventWriterT', the standard implementation of
+-- 'EventWriter'.
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE Rank2Types #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+#ifdef USE_REFLEX_OPTIMIZER
+{-# OPTIONS_GHC -fplugin=Reflex.Optimizer #-}
+#endif
+module Reflex.EventWriter
+  ( EventWriterT (..)
+  , runEventWriterT
+  , EventWriter(..)
+  ) where
+
+import Reflex.Class
+import Reflex.Host.Class
+import Reflex.PerformEvent.Class
+import Reflex.PostBuild.Class
+import Reflex.TriggerEvent.Class
+
+import Control.Monad.Exception
+import Control.Monad.Identity
+import Control.Monad.Reader
+import Control.Monad.Ref
+import Control.Monad.State.Strict
+import Data.Dependent.Map (DMap, DSum (..))
+import qualified Data.Dependent.Map as DMap
+import Data.Functor.Misc
+import qualified Data.List.NonEmpty as NonEmpty
+import Data.Maybe
+import Data.Semigroup
+import Data.Some (Some)
+import qualified Data.Some as Some
+import Data.Tuple
+import Data.Unique.Tag
+
+-- | 'MonadEventWriter' efficiently collects 'Event' values using 'tellEvent'
+-- and combines them monoidally to provide an 'Event' result.
+class (Monad m, Monoid w) => EventWriter t w m | m -> t w where
+  tellEvent :: Event t w -> m ()
+
+-- | A basic implementation of 'EventWriter'.
+newtype EventWriterT t w m a = EventWriterT { unEventWriterT :: StateT [Event t w] m a }
+  deriving (Functor, Applicative, Monad, MonadFix, MonadIO, MonadException)
+
+-- | Run a 'EventWriterT' action.
+runEventWriterT
+    :: (Reflex t, Monad m, Semigroup w)
+    => EventWriterT t w m a
+    -> m (a, Event t w)
+runEventWriterT (EventWriterT a) = do
+  (result, requests) <- runStateT a mempty
+  return (result, mconcat $ reverse requests)
```

Comment:
Except, for Events, mconcat is *far* more performant than a series of mappends.  Using Seq or something would make sense.

To answer @spl, yes, I think the reverse is important, so that the monoid is used in the same way that it is with WriterT.  I wouldn't want someone to end up surprised by things coming out in the reverse order!